test/test_struct.py

```diff
@@ -0,0 +1,172 @@
+
+from collections import OrderedDict
+from functools import partial
+import struct
+import types
+import re
+
+from six import binary_type
+
+from Xlib.protocol import rq
+from . import DummyDisplay, TestCase
+
+
+dummy_display = DummyDisplay()
+
+def pack(format, *args):
+    return struct.pack('=' + format, *args)
+
+def packstr(s, padding=0):
+    return s.encode() + (b'\0' * padding)
+
+
+class StructTest(object):
+    """ Test class helper for rq.Struct packing/unpacking support. """
+
+    struct = None
+    values = None
+    binary = None
+
+    def test_pack_value_dict(self):
+        self.assertBinaryEqual(self.struct.pack_value(dict(self.values)), self.binary)
+
+    def test_pack_value_tuple(self):
+        self.assertBinaryEqual(self.struct.pack_value(tuple(self.values.values())), self.binary)
+
+    def test_to_binary_args(self):
+        self.assertBinaryEqual(self.struct.to_binary(*self.values.values()), self.binary)
+
+    def test_to_binary_kwargs(self):
+        self.assertBinaryEqual(self.struct.to_binary(**dict(self.values)), self.binary)
+
+    def test_parse_binary(self):
+        values, remain = self.struct.parse_binary(self.binary, dummy_display)
+        self.assertEqual(values, rq.DictWrapper(dict(self.values)))
+        self.assertBinaryEmpty(remain)
+
+
+def _struct_test(name, fields):
+    """ Generate a new Struct packing/unpacking test.
+
+    <fields> is a list of (field_name, field_type, field_value, field_binary):
+        - field_name:
+            name of the field, None if it's not a named field (e.g. LengthOf)
+        - field_type:
+            factory for the field type, taking one parameter only, the name of
+            the field and returning a rq.Field subclass instance
+        - field_value:
+            the raw value passed in when packing and expected output on unpacking
+        - field_binary:
+            the binary representation of the field value or a function to
+            convert it to binary for creating the expected result of packing
+            the struct
+    """
+
+    class_name = ''.join([part.capitalize() for part in re.sub(r'[^\w]+', ' ', name).split()])
+    class_name += 'StructTest'
+    struct_layout = []
+    values = OrderedDict()
+    binary = b''
+    for field_name, field_type, field_value, field_binary in fields:
+        struct_layout.append(field_type(field_name))
+        if field_name is not None and field_value is not None:
+            values[field_name] = field_value
+        if isinstance(field_binary, binary_type):
+            binary += field_binary
+        elif isinstance(field_binary, (types.FunctionType, types.LambdaType, partial)):
+            binary += field_binary(field_value)
+        else:
+            raise ValueError('unsupported type for binary: {data} [{type}]'.format(data=str(field_binary), type=type(field_binary)))
```

Comment:
It's better to fit lines in 96 symbols as Landscape.io recommends. Unfortunately Landscape.io workers seem [down for now](https://github.com/landscapeio/landscape-issues/issues/235). For example, a long tuple can take 2 lines. I appreciate how short the code is. Sorry, if I'm too annoying with minor warnings.